services/src/main/java/org/keycloak/protocol/oidc/TokenManager.java

```diff
@@ -276,6 +283,25 @@ public boolean checkTokenValidForIntrospection(KeycloakSession session, RealmMod
         return valid;
     }
 
+    private boolean checkRevokedToken(KeycloakSession session, RealmModel realm, AccessToken token) {
+        UserSessionModel userSession = null;
+        if (token.getType().equals(TokenUtil.TOKEN_TYPE_REFRESH)) {
+            userSession = session.sessions().getUserSession(realm, token.getSessionState());
+        } else {
+            UserSessionManager sessionManager = new UserSessionManager(session);
+            userSession = sessionManager.findOfflineUserSession(realm, token.getSessionState());
+        }
+
+        ClientModel client = realm.getClientByClientId(token.getIssuedFor());
+        AuthenticatedClientSessionModel clientSession = userSession.getAuthenticatedClientSessionByClient(client.getId());
```

Comment:
Few points:
- It will be good to check if `clientSession.getCurrentRefreshToken()` matches with the current refresh token? As otherwise it may be still possible to have the scenario when incorrect refreshToken return `active: true` even if it should not.
- Another example: I am not sure if it can happen that the correct refresh token returns `active: false` even if it should be considered active.

Some example scenario to demonstrate these possible issues:
- Max Reuse is set to 2
- Refresh token A is issued from the authentication
- Refresh with Refresh Token A happens. So `clientSession.currentRefreshToken` is set to A and `clientSession.currentRefreshTokenMaxReuse` is set to 1. New refresh token B is issued
- Another refresh with refresh token A happens. So `clientSession.currentRefreshToken` is still A and `clientSession.currentRefreshTokenMaxReuse` is set to 0. New refresh token C is issued
- Introspection with C is called. I think it should be successful as token refresh would be successful as well. But it looks to me, that introspection won't be successful with your changes?
- Now refresh with token C is called. `clientSession.currentRefreshToken` will be updated to C and `lientSession.currentRefreshTokenMaxReuse` will be updated to 1.
- Introspection with token A is called. I think it should not be successful anymore, however I think that with your changes it will be successful (due the fact that clientSession.currentRefreshTokenMaxReuse is set to 1.

I think it will be good to cover the test scenario I added above.

Also I think that it will be good if the code for check of valid refresh token during introspection is not duplicated, but it re-uses same method, which is also used by the refresh token. Which is method `TokenManager.validateTokenReuse` . This will make sure that token refresh and token introspection will really use same algorithm and will be in sync with each other. Does it makes sense to you?

Reply:
@mposolda 
> Few points:
> 
> * It will be good to check if `clientSession.getCurrentRefreshToken()` matches with the current refresh token? As otherwise it may be still possible to have the scenario when incorrect refreshToken return `active: true` even if it should not.
> * Another example: I am not sure if it can happen that the correct refresh token returns `active: false` even if it should be considered active.
> 
> Some example scenario to demonstrate these possible issues:
> 
> * Max Reuse is set to 2
> * Refresh token A is issued from the authentication
> * Refresh with Refresh Token A happens. So `clientSession.currentRefreshToken` is set to A and `clientSession.currentRefreshTokenMaxReuse` is set to 1. New refresh token B is issued
> * Another refresh with refresh token A happens. So `clientSession.currentRefreshToken` is still A and `clientSession.currentRefreshTokenMaxReuse` is set to 0. New refresh token C is issued
> * Introspection with C is called. I think it should be successful as token refresh would be successful as well. But it looks to me, that introspection won't be successful with your changes?
> * Now refresh with token C is called. `clientSession.currentRefreshToken` will be updated to C and `lientSession.currentRefreshTokenMaxReuse` will be updated to 1.
> * Introspection with token A is called. I think it should not be successful anymore, however I think that with your changes it will be successful (due the fact that clientSession.currentRefreshTokenMaxReuse is set to 1.
> 
> I think it will be good to cover the test scenario I added above.

I see. Thank you for your explaination.
 
> Also I think that it will be good if the code for check of valid refresh token during introspection is not duplicated, but it re-uses same method, which is also used by the refresh token. Which is method `TokenManager.validateTokenReuse` . This will make sure that token refresh and token introspection will really use same algorithm and will be in sync with each other. Does it makes sense to you?

I see. I want to modify as the following.

- Create a new method
- The common process that token refresh and token introspection should share move from `TokenManager.validateTokenReuse()` to the new method.
- The new method is called from `TokenManager.validateTokenReuse()` and `TokenManager.checkRevokedToken()`.

What do you think?

Reply:
@Michito-Okai Yes, I think that something like you propose makes sense. Also it will be nice to possibly rename existing methods on TokenManager to be more descriptive like `validateTokenReuseForIntrospection` and `validateTokenReuseForRefresh` and make them call the new method you propose

Reply:
@mposolda 
I fixed. Could you confirm it?